sdk/python/lib/pulumi/_types.py

```diff
@@ -657,6 +657,13 @@ def get_origin(tp):
         return None
 
     def get_args(tp):
+        # Emulate the behavior of get_args for Union on Python 3.6.
```

Comment:
Basically, what was happening was that `tp.__args__` for `Union` wasn't returning the flattened list of args on Python 3.6 like later versions. So when we were [inspecting those args](https://github.com/pulumi/pulumi/blob/69d50ced27920c30d519313603cb247654ea7341/sdk/python/lib/pulumi/_types.py#L906-L933) to determine if the type annotation was intended to be an input class vs. user-defined dict, we weren't able to because the args we were getting weren't what was expected.

We have tests for this https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/test/test_types_input_type_types.py that were failing on Python 3.6 before the change in this PR, but we weren't running the tests against Python 3.6 in CI.

Reply:
Yup. Reproduced fail/fix on Py 3.6. This is the substitution that _eval is doing (on `pulumi.Input[str]` context):

```
                assert _eval((str, (typing.Awaitable, str), (pulumi.output.Output, None))) == \
                    (str, typing.Awaitable[str], pulumi.output.Output[None])
```